nats/client.go

```diff
@@ -430,103 +420,96 @@ func (c *Client) initObjectStore(ctx context.Context, accountId string) error {
-	if err != nil {
-		return err
-	}
-
-	// Publish the source message with replayed sequence ID so it's picked up by
-	// ephemeral consumer
-	c.Publish(ctx, rawMsg.Data, ChannelNotify, replaySequenceId, "event")
-
-	// Set the consumer on the client
-	c.Consumer = consumer
-	return nil
-}
-
-func (c *Client) initWorkerConsumer(ctx context.Context, accountId string, appName string) error {
-	name := fmt.Sprintf("%s-%s-%s", accountId, ChannelRequest, appName)
-	// Create or update the consumer, since these are created dynamically
-	consumerCfg := jetstream.ConsumerConfig{
-		Name:          name,
-		Durable:       name,
-		FilterSubject: WorkerRequestSubject(accountId, appName, "*"),
-		AckWait:       1 * time.Minute,
-	}
-	consumer, err := c.JetStream.CreateOrUpdateConsumer(ctx, accountId, consumerCfg)
-	if err != nil {
-		return err
-	}
-
-	c.Consumer = consumer
-	return nil
-}
-
 // ClientOpts - passed through to NewClient() to configure the client setup
 
 // DefaultClientOpts configures the hiphops nats.Client as a RunnerClient
 func DefaultClientOpts() []ClientOpt {
 	return []ClientOpt{
-		RunnerClient(),
+		RunnerClient(DefaultConsumerName),
 	}
 }
 
 // ReplayClient initialises the client with a consumer for replaying a sequence
-func ReplayClient(sequenceId string) ClientOpt {
+func ReplayClient(name string, sequenceId string) ClientOpt {
 	return func(c *Client) error {
-		err := c.initReplayConsumer(context.Background(), c.accountId, sequenceId)
+		ctx := context.Background() // TODO: Move all context creation in ClientOpts to argument rather than in function
+
+		// Get the source message from the stream
+		stream, err := c.JetStream.Stream(ctx, c.accountId)
 		if err != nil {
 			return err
 		}
+
+		// Get the source message to be replayed from the stream
+		sourceMsgSubject := SourceEventSubject(c.accountId, sequenceId)
+		rawMsg, err := stream.GetLastMsgForSubject(ctx, sourceMsgSubject)
+		if err != nil {
+			return fmt.Errorf("Failed to fetch source event: %w", err)
+		}
+		if rawMsg == nil {
+			return fmt.Errorf("No source event found for subject '%s'", sourceMsgSubject)
+		}
+
+		// Create a new, random replay sequence ID
+		replaySequenceId := fmt.Sprintf("replay-%s", uuid.NewString()[:20])
+
+		// Create ephemeral consumer filtered by replayed sequence ID
+		consumerCfg := jetstream.ConsumerConfig{
+			Name:          replaySequenceId,
+			Description:   fmt.Sprintf("Replay request for sequence: '%s'", sequenceId),
+			FilterSubject: ReplayFilterSubject(c.accountId, replaySequenceId),
+			DeliverPolicy: jetstream.DeliverAllPolicy,
+		}
+		consumer, err := c.JetStream.CreateConsumer(ctx, c.accountId, consumerCfg)
+		if err != nil {
+			return err
+		}
+
+		// Publish the source message with replayed sequence ID so it's picked up by
+		// ephemeral consumer
+		c.Publish(ctx, rawMsg.Data, ChannelNotify, replaySequenceId, "event")
+
+		// Set the consumer on the client
+		c.Consumers[name] = consumer
```

Comment:
When does this ephemeral consumer get cleaned up? Does it need to be? (Otherwise I think this array is unbounded).

Reply:
Sorry, not sure what you mean by 'I think this array is unbounded'?

On the consumer side, ephemeral consumers are automatically cleaned up by nats. We don't need to clean them up on our side as they're just references to a resource in nats. We do clean up consumer contexts, but that happens elsewhere and seems to be fine.

Reply:
I mean, this map `c.Consumers` gets a new entry on every ephemeral consumer. Does that entry ever get removed?